pinot-controller/src/main/java/com/linkedin/pinot/controller/api/resources/PinotSegmentUploadRestletResource.java

```diff
@@ -498,31 +479,26 @@ private PinotResourceManagerResponse uploadSegment(File indexDir, SegmentMetadat
       existingSegmentZKMetadata.setCustomMap(
           segmentZKMetadataCustomMapModifier.modifyMap(existingSegmentZKMetadata.getCustomMap()));
 
-      PinotResourceManagerResponse response;
+      // Update ZK metadata and refresh the segment if necessary
       long newCrc = Long.valueOf(segmentMetadata.getCrc());
-      if (newCrc != existingCrc) {
-        if (downloadUrl == null) {
-          downloadUrl = moveSegmentToPermanentDirectory(provider, tableName, segmentName, tempTarredSegmentFile);
+      if (newCrc == existingCrc) {
+        // New segment is the same as the existing one, only update ZK metadata without refresh the segment
+        if (!_pinotHelixResourceManager.updateZkMetadata(existingSegmentZKMetadata)) {
+          throw new RuntimeException(
+              "Failed to update ZK metadata for segment: " + segmentName + " of table: " + offlineTableName);
         }
-        // If refreshSegment returns without an exception, it should have already written to zk
-        response = _pinotHelixResourceManager.refreshSegment(segmentMetadata, existingSegmentZKMetadata, downloadUrl);
       } else {
-        // New segment is the same as the existing one, only update ZK metadata without refresh the segment
-        _pinotHelixResourceManager.updateZkMetadata(existingSegmentZKMetadata);
-        response = PinotResourceManagerResponse.SUCCESS_RESPONSE;
-      }
-      metadataRewritten = true;
-
-      if (!response.isSuccessful()) {
-        _controllerMetrics.addMeteredGlobalValue(ControllerMeter.CONTROLLER_SEGMENT_UPLOAD_ERROR, 1L);
-        throw new ControllerApplicationException(LOGGER, "Error uploading segment",
-            Response.Status.INTERNAL_SERVER_ERROR);
+        // New segment is different with the existing one, update ZK metadata and refresh the segment
+        if (downloadUrl == null) {
+          downloadUrl = moveSegmentToPermanentDirectory(provider, rawTableName, segmentName, tempTarredSegmentFile);
+        }
+        _pinotHelixResourceManager.refreshSegment(segmentMetadata, existingSegmentZKMetadata, downloadUrl);
       }
-      return response;
-    } finally {
-      if (enableParallelPushProtection && !metadataRewritten) {
-        _pinotHelixResourceManager.updateZkMetadata(existingSegmentZKMetadata);
+    } catch (Exception e) {
+      if (!_pinotHelixResourceManager.updateZkMetadata(existingSegmentZKMetadata)) {
```

Comment:
i could be missing something, but why do we not need to reset segment upload time upon exceptions thrown?

Reply:
We need to, and this statement is resetting the upload time (see line 464)